ext/curl/share.c

```diff
@@ -134,6 +136,112 @@ PHP_FUNCTION(curl_share_strerror)
 }
 /* }}} */
 
+/**
+ * Initialize a persistent curl share handle with the given share options, reusing an existing one if found.
+ *
+ * Throws an exception if the share options are invalid.
+ */
+PHP_FUNCTION(curl_persistent_share_init)
+{
+	zval *share_opts = NULL, *entry = NULL;
+	zend_ulong persistent_id = 0;
+
+	php_curlsh *sh;
+
+	CURLSHcode error;
+
+	ZEND_PARSE_PARAMETERS_START(1, 1)
+		Z_PARAM_ARRAY_EX(share_opts, 0, 1)
+	ZEND_PARSE_PARAMETERS_END();
+
+	object_init_ex(return_value, curl_persistent_share_ce);
+	sh = Z_CURL_SHARE_P(return_value);
+
+	ZEND_HASH_FOREACH_VAL(Z_ARRVAL_P(share_opts), entry) {
+		ZVAL_DEREF(entry);
+
+		zend_ulong option = zval_get_long_ex(entry, true);
+
+		if (option == CURL_LOCK_DATA_COOKIE) {
+			zend_throw_exception_ex(
+				NULL,
+				0,
+				"CURL_LOCK_DATA_COOKIE is not allowed with persistent curl share handles"
+			);
+
+			goto error;
+		}
+
+		// Ensure that each additional option results in a unique persistent ID.
+		persistent_id += 1 << option;
```

Comment:
Is it guaranteed that all the option values are below 32 / 64? Otherwise this results in undefined behavior.

Also you should likely use `|=`, not `+=` to protect against redundantly providing the same option twice.

Reply:
> Is it guaranteed that all the option values are below 32 / 64? Otherwise this results in undefined behavior.

They currently are (though the user could specify an invalid option value which is included in the persistent ID before it's checked): https://github.com/curl/curl/blob/72edb22b8a0e1e62216b59a045147d1a863ad59b/include/curl/curl.h#L2993-L3008. If I switch to a strict allow-list as you suggested above, I believe that would solve both potential sources of an option > 32?

> Also you should likely use |=, not += to protect against redundantly providing the same option twice.

Good catch, will change.

Reply:
> If I switch to a strict allow-list as you suggested above, I believe that would solve both potential sources of an option > 32?

Yes, that's why I suggested it.

Reply:
I've used this suggestion in the latest commit.